src/kmp.h

```diff
@@ -0,0 +1,68 @@
+#ifndef KMP_H
+#define KMP_H
+
+#include <malloc.h>
+
+int *compute_border_array(char *pattern, int pattern_len) {
+    if (pattern_len == 0) return NULL;
+    int *ba = malloc(sizeof(int)*pattern_len);
+    ba[0] = 0;
+    for (int i = 1; i < pattern_len; i++) {
+        // Test if we can extend border. Otherwise, reset.
+        if (pattern[i] == pattern[ba[i-1]]) {
+            ba[i] = ba[i-1]+1;
+        } else {
+            ba[i] = 0;
+        }
+    }
+    return ba;
+}
+
+int *compute_restricted_border_array(char *pattern, int pattern_len) {
+    // Compute and change the border array.
+    // Could be done faster (and uglier), but we assume pattern_len is not long.
+    int *bax = compute_border_array(pattern, pattern_len);
+    for (int i = 1; i < pattern_len; i++) {
+        if (pattern[i+1] == pattern[bax[i]]) {
+            // Next char after border must be different
+            bax[i] = bax[bax[i]-1];
+        }
+    }
+    return bax;
+}
+
+// Checks sequence[seq_index:] for pattern. Returns index when exact pattern is found. Otherwise -1.
+int kmp_get_position_of_next_match(char *sequence, char *pattern, int seq_index, int pattern_len, int *bax) {
+    // We define, that the empty string can be found everywhere
+    if (pattern_len == 0) return seq_index+1;
+    int seq_len = (int) strlen(sequence);
+
+    int i = seq_index;
+    int j = 0;
+
+    while (i <= seq_len-pattern_len+j) {
+        // Keep increasing i and j on matches
+        while (j < pattern_len && sequence[i] == pattern[j]) { i++; j++; }
+
+        // If we have matched all the way; return.
+        if (j == pattern_len) return i-pattern_len+1;
+
+        // On mismatches, we can use bax to skip
+        if (sequence[i] != pattern[j]) {
+            if (j == 0) {
+                i++;
+            } else {
+                j = bax[j - 1];
+            }
+        }
+    }
+    return -1;
+}
+
+// Calls the original function with a precomputed restricted border array
+int kmp_get_position_of_next_match_testing(char *sequence, char *pattern, int seq_index, int pattern_len) {
+    int *bax = compute_restricted_border_array(pattern, pattern_len);
+    return kmp_get_position_of_next_match(sequence, pattern, seq_index, pattern_len, bax);
+}
```

Comment:
![no-god-please-no-no](https://user-images.githubusercontent.com/3206538/193411935-77bf1989-b62a-431e-9712-633a464c6c3a.gif)

If this is something you plan to call with multiple `seq_index`--and I really hope you are not planning that--then this is *obviously* a bad idea to allow a user access to a function that will compute the border array in $O(m)$ for each of the $n$ positions in `x`. Right?

The running time of `strlen()` I can't expect you to know, but you bleeping know that it takes time $O(m)$ to compute the border array...

Also, what do you think happens with the memory you allocate for `bax`? I know that `compute_restricted_border_array()` allocates memory. I know that `kmp_get_position_of_next_match()` doesn't free it. So every time you call this function, you leak `m` integers. Forget about the running time, you won't have the memory to run this on anything serious.

Please delete this function before someone calls it.

Reply:
Sorry, this function should have been in test/tests.c. You are right; storing it in kmp.h conveys a wrong meaning. The function is only used for testing purposes.

tests.c tests exact pattern matching for the naive and kmp implementations with a "test function":
int (*test_fun)(char*, char*, int, int);

That is, a function with the above signature, is something that we have for both implementations, which is supposed to return the next match in the string. We then set the test_fun and run all tests - see below:

```
    // Run tests for the naive impl.
    test_fun = &naivly_get_position_of_next_match;
    run_all_tests();

    // Run tests for the linear impl.
    test_fun = &kmp_get_position_of_next_match_testing;
    run_kmp_relevant_tests();
    run_all_tests();
```

Reply:
Using iterators is a great idea. There is nothing wrong with that. And sometimes, you can get away with using a simple function interface for them, like having a `next(x, pos, p, whatever)` call. But sometimes your algorithms do not need the same information, which is the case here where the naive algorithm only needs the two strings and a position to start from (or simpler still when you remember that a string is just a nul-terminated buffer, and thus you can get away with just giving it two `char *` pointers). For KMP, however, you also need the border array or jump table, and you need to remember not just the position in `x` but also where you are in `p`.

The solution is to wrap up what you need in a structure. The simplest approach is to make an iterator interface that gets the extra data through a `void *` pointer.

```c
// interface
typedef int (*search_iterator)(const char *x, int i, const char *p, void *data);

// implementation
int naive(const char *x, int i, const char *p, void *)
{
   // search in x without needing any data
}

int kmp(const char *x, int i, const char *p, void *data)
{
   // get the kmp-specific data
   struct kmp_data *kmp_data = data;
   int *ba = kmp_data->ba;
   int j = kmp_data->j;
   // search for the next occurrence.
   // before returning, remember to update kmp_data->j
}
```

This is not particularly type-safe, though. Whenever you need to go through a `void *` you toss away all type-information. Those can point to any data type, and any data type pointer can be assigned to `void *` without any type check. That makes it easy to call the wrong function with the wrong type, and you won't know until the program crashes. Hopefully, it does quickly, but it can crash in mysterious ways long after you made the mistake.

A better solution is to wrap up iterators as polymorphic types. Here, the problem is that C doesn't support polymorphism, but that should not stop us. We can define a generic iterator type using function pointers for dynamic dispatch. (This can cost a little in performance since those can mess with branch prediction, but if it is not in a tight loop, you will never notice).

Here's an example I wrote a little while back:

```c
typedef struct
{ // Embed this in matchers (with the right v-table)
  struct cstr_exact_matcher_vtab *vtab;
} cstr_exact_matcher;


typedef struct cstr_exact_matcher_vtab
{ // Virtual table for dynamic dispatching
  // Implements iteration
  long long (*next)(cstr_exact_matcher *);
  // Implements destruction
  void (*free)(cstr_exact_matcher *);
} cstr_exact_matcher_vtab;


// returns -1 when there are no more matches, otherwise an index of a match
INLINE long long cstr_exact_next_match(cstr_exact_matcher *self)   { return self->vtab->next(self); }
INLINE void      cstr_free_exact_matcher(cstr_exact_matcher *self) { self->vtab->free(self); }
```

https://github.com/mailund/cstr/blob/08ef7895aa9e6c2dbae1cb457369522773529ac1/src/cstr.h#L480-L497

There is an abstract type, `struct cstr_exact_matcher`, which contains a pointer to a virtual function table (vtable) with the functions I should be able to call on an iterator. You could embed the function pointers in the iterator instead, but for each type, you will have the same functions, so for more complex types, you generally want a vtable per type and for each instance to point at this table, which is what I do here.

The vtable here only has two functions: one for finding the next match (like you have) and one for freeing an iterator (something you should have since you are leaking memory).

For calling the functions, you need to get the function in the correct vtable (that you get from the object), and you want the tables to take the object itself as arguments. Perhaps you need more arguments, but you need at least those (and they are all I needed here). Write functions that do the vtable lookup and call the functions; that is the easiest way to ensure you get it right each time.

Then, if you have multiple types of iterators, you can make constructors for them:

```c
cstr_exact_matcher *cstr_naive_matcher(cstr_const_sslice x, cstr_const_sslice p);
cstr_exact_matcher *cstr_ba_matcher(cstr_const_sslice x, cstr_const_sslice p);
cstr_exact_matcher *cstr_kmp_matcher(cstr_const_sslice x, cstr_const_sslice p);
```

In this implementation, I make the iterators hold on to the strings they are supposed to iterate over. Nothing good will come of trying to switch the data in the middle of an iteration, so I don't need to change that, and I would rather that the user didn't make it by mistake. That is why the two vtable functions only take the iterator as their argument; I don't need anything more, as I already have the strings in the iterator.

For the vtable stuff to work, you have to ensure that all the concrete types have a vtable where we can find it, and that means that the initial elements in the structures have to be the same. C is simple, and it will just look at memory offsets when you get elements in a struct, but that also means that if we put what we need in the same order in different structs, then the C standard ensures that they will be at the same offsets, and that means that we can cast and get what we need.

Generally, you just want the vtable pointer at the top of the concrete objects. (If you want inheritance, then you also need to nest tables correctly, but that is not necessary here). I also want the structures to always hold `x` and `p`, so I have defined the shared bits as this:

```c
#define SHARED                                        \
    cstr_exact_matcher matcher; /* MUST come first */ \
    cstr_const_sslice x;                              \
    cstr_const_sslice p;
```
https://github.com/mailund/cstr/blob/08ef7895aa9e6c2dbae1cb457369522773529ac1/src/exact.c#L10-L13

and I put that at the top of all the concrete s structures.

```c
struct naive_matcher_state
{
    SHARED
    long long i;
};
```
https://github.com/mailund/cstr/blob/08ef7895aa9e6c2dbae1cb457369522773529ac1/src/exact.c#L33-L37

```c
struct ba_matcher_state
{
    SHARED
    long long i, b;
    long long ba[];
};
```
https://github.com/mailund/cstr/blob/08ef7895aa9e6c2dbae1cb457369522773529ac1/src/exact.c#L95-L100

```c
struct kmp_matcher_state
{
    SHARED
    long long i, j;
    long long ba[];
};
```
https://github.com/mailund/cstr/blob/08ef7895aa9e6c2dbae1cb457369522773529ac1/src/exact.c#L142-L147

When you want to create an object, you need to set the vtable pointer. You can create one such vtable per type statically.

```c
// Helper macro for initialising matcher v-tables
#define MATCHER_VTAB(NEXT, FREE)   \
    .next = (exact_next_fn)(NEXT), \
    .free = (exact_free_fn)(FREE),
```
https://github.com/mailund/cstr/blob/08ef7895aa9e6c2dbae1cb457369522773529ac1/src/exact.c#L21-L24

```c
static cstr_exact_matcher_vtab naive_vtab = {MATCHER_VTAB(naive_next, free)};
static cstr_exact_matcher_vtab ba_vtab = {MATCHER_VTAB(ba_next, free)};
static cstr_exact_matcher_vtab kmp_vtab = {MATCHER_VTAB(kmp_next, free)};
```

That connects the types and the right search functions, such that once you have created an iterator (if you connect the vtable pointer correctly) it will always use the right function.

When you create objects, point to the right vtable

```c
// Initialising shared bits
#define MATCHER(VTAB, X, P)       \
    .matcher = {.vtab = &(VTAB)}, \
    .x = (X),                     \
    .p = (P)
```
https://github.com/mailund/cstr/blob/08ef7895aa9e6c2dbae1cb457369522773529ac1/src/exact.c#L15-L19

```c
cstr_exact_matcher *cstr_naive_matcher(cstr_const_sslice x, cstr_const_sslice p)
{
    struct naive_matcher_state *state = cstr_malloc(sizeof *state);
    *state = (struct naive_matcher_state){
        MATCHER(naive_vtab, x, p), .i = 0};
    return (cstr_exact_matcher *)state;
}

cstr_exact_matcher *cstr_ba_matcher(cstr_const_sslice x, cstr_const_sslice p)
{
    // allocate space for the the struct + the border array
    // in the flexible array for ba.
    struct ba_matcher_state *state =
        CSTR_MALLOC_FLEX_ARRAY(state, ba, (size_t)p.len);
    *state = (struct ba_matcher_state){
        MATCHER(ba_vtab, x, p), .i = 0, .b = 0};
    compute_border_array(p, state->ba);
    return (cstr_exact_matcher *)state;
}

cstr_exact_matcher *cstr_kmp_matcher(cstr_const_sslice x, cstr_const_sslice p)
{
    struct kmp_matcher_state *state =
        CSTR_MALLOC_FLEX_ARRAY(state, ba, (size_t)p.len);
    *state = (struct kmp_matcher_state){
        MATCHER(kmp_vtab, x, p),
        .i = 0, .j = 0};
    compute_border_array(p, state->ba);
    return (cstr_exact_matcher *)state;
}
```

You probably don't need things to be quite this complicated (but really, once you have written such code a few times, it doesn't feel that complicated any longer), but you do need to provide data to the functions and for that, putting data in the iterators is the way to go. It is now what the standard functions in C do, but they are quite simple and don't need to. If you only need very simple things, then this is overkill, but that is not the world you are in right here.